tsl/src/remote/dist_txn.c

```diff
@@ -20,16 +21,56 @@
 
 #ifdef DEBUG
 
-void (*testing_callback_call_hook)(const char *event) = NULL;
+static const DistTransactionEventHandler *event_handler = NULL;
+static const char *eventnames[MAX_DTXN_EVENT] = {
+	[DTXN_EVENT_ANY] = "any",
+	[DTXN_EVENT_PRE_COMMIT] = "pre-commit",
+	[DTXN_EVENT_WAIT_COMMIT] = "waiting-commit",
+	[DTXN_EVENT_PRE_ABORT] = "pre-abort",
+	[DTXN_EVENT_PRE_PREPARE] = "pre-prepare-transaction",
+	[DTXN_EVENT_WAIT_PREPARE] = "waiting-prepare-transaction",
+	[DTXN_EVENT_POST_PREPARE] = "post-prepare-transaction",
+	[DTXN_EVENT_PRE_COMMIT_PREPARED] = "pre-commit-prepared",
+	[DTXN_EVENT_WAIT_COMMIT_PREPARED] = "waiting-commit-prepared",
+	[DTXN_EVENT_SUB_XACT_ABORT] = "subxact-abort",
+};
+
+void
+remote_dist_txn_set_event_handler(const DistTransactionEventHandler *handler)
+{
+	event_handler = handler;
+}
+
+static inline void
+eventcallback(const DistTransactionEvent event)
+{
+	if (NULL != event_handler && NULL != event_handler->handler)
+		event_handler->handler(event, event_handler->data);
+}
+
+DistTransactionEvent
+remote_dist_txn_event_from_name(const char *eventname)
+{
+	int i;
+
+	for (i = 0; i < MAX_DTXN_EVENT; i++)
+	{
+		if (strcmp(eventname, eventnames[i]) == 0)
+			return i;
+	}
+
+	elog(ERROR, "invalid event name");
+	pg_unreachable();
+}
+
+const char *
+remote_dist_txn_event_name(const DistTransactionEvent event)
+{
```

Comment:
Maybe something like assert `event >= 0 && event < MAX_DTXN_EVENT` ?

Reply:
even is an enum so it is not possible to have values that do not correspond to a valid event.